src/doc/rustdoc/src/how-to-read-rustdoc.md

```diff
@@ -0,0 +1,105 @@
+At the top is some at-a-glance info and controls:
+
+- the type and name of the item,
+  such as "Struct `std::time::Duration`",
+- a button to copy the item's path to the clipboard,
+  which is a clipboard item
+- a button to collapse or expand the top-level documentation for that item
+  (`[+]` or `[-]`),
+- a link to the source code (`[src]`),
+  if [configured](the-doc-attribute.html#html_no_source),
+- and the version in which the item became stable,
+  if it's a stable item in the standard library.
+
+Below this is the main documentation for the item,
+including a definition or function signature if appropriate,
+followed by a list of fields or variants for Rust types.
+Finally, the page lists associated functions and trait implementations,
+including automatic and blanket implementations that `rustdoc` knows about.
+
+### Navigation
+
+Subheadings, variants, fields, and many other things in this documentation
+are anchors and can be clicked on and deep-linked to,
+which is a great way to communicate exactly what you're talking about.
+The typograpical character "§" appears next to lines with anchors on them
+when hovered or given keyboard focus.
+
+## The Navigation Bar
+
+For example, when looking at documentation for the crate root,
+it shows all the crates documented in the documentation bundle,
+and quick links to the modules, structs, traits, functions, and macros available
+from the current crate.
+At the top, it displays a [configurable logo](the-doc-attribute.html#html_logo_url)
+alongside the current crate's name and version,
+or the current item whose documentation is being displayed.
+
+## The Theme Picker and Search Interface
+
+When viewing `rustdoc`'s output in a browser with JavaScript enabled,
+a dynamic interface appears at the top of the page.
+To the left is the theme picker, denoted with a paint-brush icon,
+and the search interface, help screen, and options appear to the right of that.
+
+### The Theme Picker
+
+Clicking on the theme picker provides a list of themes -
+by default `ayu`, `light`, and `dark` -
+which are available for viewing.
+
+### The Search Interface
+
+Typing in the search bar instantly searches the available documentation for
+the string entered with a fuzzy matching algorithm that is tolerant of minor
+typos.
+
+By default, the search results give are "In Names",
+meaning that the fuzzy match is made against the names of items.
+Matching names are shown on the left, and the first few words of their
+descriptions are given on the right.
+By clicking an item, you will navigate to its particular documentation.
+
+There are two other sets of results, shown as tabs in the search results pane.
+"In Parameters" shows matches for the string in the types of parameters to
+functions, and "In Return Types" shows matches in the return types of functions.
+Both are very useful when looking for a function whose name you can't quite
+bring to mind when you know the type you have or want.
```

Comment:
It's also useful if you don't know whether the function exists, but you think it should (e.g. `Option<T> -> T` or `String -> PathBuf`).

Reply:
Just noting that neither of those examples work as search queries currently. I assume you didn't mean using those as queries verbatim?

Reply:
Sure it does: 
![image](https://user-images.githubusercontent.com/23638587/139919466-0276152d-b271-4c90-805a-5c0936f51313.png)
 https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html#method.from-5

Reply:
The results include functions that don't have `String` in their parameters (like `Path::join`). `String -> PathBuf` implies that it will show only functions that takes `String`s and return `PathBuf`s.

Reply:
:man_shrugging: I don't think there exist any useful type-based search queries for which rustdoc doesn't have bugs. I don't think we should block documenting how to use the feature on fixing the bugs.

Reply:
I just don't think we should give examples like `String -> PathBuf` since they won't behave as expected. We could suggest `-> PathBuf` or `* -> PathBuf` instead though (not sure which one of those is the more "correct" way).

Reply:
Those aren't *useful* searches though. We should give an example of why you would want to use the feature.

Reply:
What do you mean? `String -> PathBuf` isn't useful either, because it returns the same results. Personally, I think we should avoid documenting type-based search here until the dust settles a bit (since Guillaume is working on an overhaul of it right now). After Guillaume finishes his overhaul, IIUC `String -> PathBuf` should work correctly.

> We should give an example of why you would want to use the feature.

Giving a broken example of why you'd want to use the feature seems worse than giving no example at all.

Reply:
Yeah, I think we should wait on this until it's updated, if that's already in progress.

Reply:
(I opened the PR [here](https://github.com/rust-lang/rust/pull/90630)).